lib/pbench/server/database/models/user.py

```diff
@@ -90,36 +98,39 @@ def evaluate_role(self, key, value):
     def evaluate_password(self, key, value):
         return generate_password_hash(value)
 
-    # validate the email field
     @validates("email")
     def evaluate_email(self, key, value):
         valid = validate_email(value)
-        email = valid.email
+        return valid.email
 
-        return email
+    def add_token(self, token: ActiveToken):
+        """Add the given token to active tokens list for this user."""
+        try:
+            self.auth_tokens.append(token)
+            Database.db_session.add(token)
+            Database.db_session.commit()
+        except Exception:
+            Database.db_session.rollback()
+            raise
```

Comment:
That's an interesting question; without experimenting I'm not quite sure how this will fall out. I see two distinct styles of inserting child rows:
1. (Which I do in metadata) is to simply set the parent link in the child and then *add* it to the session, and let `commit` set the reciprocal link (which is really just the collection of matching parent foreign key rows in the child table);
2. (Which seems to be the "canonical example") is to add the child to the parent's relationship set (note, by default a `list` but you can also actually make it a `set`), and let the `commit` set the reciprocal link.

I'm not quite sure how smart `rollback` is on this path. Will it implicitly remove the child from the parent's relationship list, or leave it?

Reply:
> Second, `self` refers to an entry in the `Users` table (right?), which is going away real soon now, so perhaps we don't need to care?

I like that answer, and if the tests pass we're likely OK regardless of any hypothetical implications for the intricate internal SQLAlchemy magic in the short term.